lib/yamatanooroti/vterm.rb

```diff
@@ -42,76 +40,121 @@ def write(str)
-    sync
+    # Write str (e.g. `exit`) to pty_input might terminate the process.
+    try_sync
   end
 
   def close
-    sync
-    @pty_input.close
-    sync
-    Process.kill('KILL', @pid)
-    Process.waitpid(@pid)
+    begin
+      sync
+      @pty_input.close
+      sync
+    rescue IOError, Errno::EIO
+    end
+    begin
+      Process.kill('KILL', @pid)
+      Process.waitpid(@pid)
+    rescue Errno::ESRCH
+    end
   end
 
-  private def sync
-    startup_message = +'' if @startup_message
+  private def wait_startup_message
+    wait_until = Time.now + @timeout
+    chunks = +''
     loop do
-      sleep @wait
-      chunk = @pty_output.read_nonblock(1024)
-      if @startup_message
-        startup_message << chunk
-        if @startup_message.(startup_message)
-          @startup_message = nil
-          chunk = startup_message
-        else
-          redo
-        end
+      wait = wait_until - Time.now
+      if wait.negative? || !@pty_output.wait_readable(wait)
+        raise "Startup message didn't arrive within timeout: #{chunks.inspect}"
       end
-      @vterm.write(chunk)
-      chunk = @vterm.read
-      @pty_input.write(chunk)
-    rescue Errno::EAGAIN, Errno::EWOULDBLOCK
-      retry if @startup_message
-      break
-    rescue Errno::EIO # EOF
-      retry if @startup_message
-      break
-    rescue IO::EAGAINWaitReadable # emtpy buffer
-      retry if @startup_message
-      break
+
+      chunk = @pty_output.read_nonblock(65536)
+      vterm_write(chunk)
+      chunks << chunk
+      break if yield chunks
     end
   end
 
+  private def vterm_write(chunk)
+    @vterm.write(chunk)
+    @pty_input.write(@vterm.read)
+    @result = nil
+  end
+
+  private def sync(wait = @wait)
+    sync_until = Time.now + @timeout
+    while @pty_output.wait_readable(wait)
+      vterm_write(@pty_output.read_nonblock(65536))
+      break if Time.now > sync_until
+    end
+  end
+
+  private def try_sync(wait = @wait)
+    sync(wait)
+    true
+  rescue IOError, Errno::EIO
+    false
+  end
+
+
   def result
-    return @result if @result
-    @result = []
+    try_sync(0)
+    @result ||= retrieve_screen
+  end
+
+  private def retrieve_screen
+    result = []
     rows, cols = @vterm.size
     rows.times do |r|
-      @result << +''
+      result << +''
       cols.times do |c|
         cell = @screen.cell_at(r, c)
         if cell.char # The second cell of fullwidth char will be nil.
           if cell.char.empty?
             # There will be no char to the left of the rendered area if moves
             # the cursor.
-            @result.last << ' '
+            result.last << ' '
           else
-            @result.last << cell.char
+            result.last << cell.char
           end
         end
       end
-      @result.last.gsub!(/ *$/, '')
+      result.last.gsub!(/ *$/, '')
+    end
+    result
+  end
+
+  private def assert_screen_with_proc(check_proc, assert_block, convert_proc = :itself.to_proc)
```

Comment:
It feels like using `assert` gives the impression of a public method for testing, so maybe a different name would be better. (Sorry, I don’t have any ideas right now.)

Reply:
Changed to `retryable_screen_assertion_with_proc`